Forestry Support

TeknoKit opened the pull request:
At the start of any isEqual comparison, if Forestry is loaded, it determines if either item is a Forestry Bee with a GEN tag and if so flags the comparison with COMPARE_STRIP_NBT. A forestry case with the GEN tag has been added to the stripTags method.
Is this better?

TeknoKit committed:
New Forestry Support Implementation

If Forestry is loaded, a new check is made when beginning to compare any items to see if any item in question is a Forestry Queen/Princess Bee, and if so, if it also has the "GEN" tag, to flag the comparison for NBT striping to remove the "GEN" tag.

way2muchnoise commented on src/main/java/com/raoulvdberge/refinedstorage/integration/forestry/IntegrationForestry.java:
```diff
@@ -0,0 +1,64 @@
+     * @return		previous compare flags with COMPARE_STRIP_NBT if any itemstack in list is a 
+     * 				Forestry Queen/Princess Bee with a GEN NBT tag, otherwise only previous compare flags
+     */
+	public static int isItem(int flags, ItemStack... items) {
```
This method is kind ambiguous as `isItem` should return a `boolean` indicating whether it is a forestry item or not. Would rather see this method replace with something like a `isBee` method that takes a single `ItemStack`.

way2muchnoise replied:
As mentioned in the other comment it would be nice if this also had support for trees and butterflies. Might as well go all with this support.

raoulvdberge commented:
No, it's not. The comparer needs to guarantee correct compares, we can't just strip nbt when not asked to strip.

way2muchnoise commented on src/main/java/com/raoulvdberge/refinedstorage/apiimpl/util/Comparer.java:
```diff
@@ -20,6 +21,11 @@ public boolean isEqual(@Nullable ItemStack left, @Nullable ItemStack right, int
+        
+        if(IntegrationForestry.isLoaded()) {
+        	ItemStack[] items = {left, right};
+        	flags = IntegrationForestry.isItem(flags, items);
```
Changing the flag here seems kinda off. I would propose adding a method in `IntegrationForestry` along the lines of `sanitise`, where it removes then tags that need to be removed without changing the comparison flags. The this if would be looking more like this:

```java
if (IntegrationForestry.isLoaded()) {
    if (IntegrationForestry.isBee(left) {
        left = IntegrationForestry.sanitise(left.copy());
    }
    if (IntegrationForestry.isBee(right)) {
        right = IntegrationForestry.sanitise(right.copy());
    }
}
```

In this approach `sanitise` could be added to or changed without changing this code.

way2muchnoise replied:
This `isBee` method could be changed to something that would also support trees and butterflies. And name it appropriately, eg. `isBreedable`.

TeknoKit commented:
I was under the impression that modifying NBT tags directly outside of comparison was not wanted and thus tried to integrate it to the one method that did such operations but wanted to avoid that for your stated reason the first time. Either way I do understand why it was better to be moved.
I am also sure that moving above the IComparer is a bad idea; trying to ensure all the different items that can extract and insert all accommodate for these items is bound to just get messy and mistakes made.

Yes, I could fairly quickly add in trees and butterflies once I understand the NBTCompounds.
I'll get this right eventually, but I did need some gauging to see the proper mindset here. I'll work on it.

TeknoKit committed:
Added more Item support framework, refactored sanitation

Found one case in which to never ingore NBT tags, which is when dealing with items inside of a grid. Otherwise items with a tag being ignored will become exactly the same as the first equal item found. Added a NO_FORESTRY_SANTIZE flag to mark these cases, and marked the insert method for StorageDIsks. This prevents the (Forestry) item from being converted into other "equal" items, but leaves GUI bugs.
Unsure if I should continue trying to flag non-sanitation instances, or to instead mark methods that should sanitize.

TeknoKit commented:
I think setting appropriate methods with a sanitize flag is better than altering the default behavior and then trying to blacklist some methods. I'll change that behavior back and put the flags in properly.

TeknoKit committed:
Changed to whitelist flag

Nothing flagged yet, as having trouble understanding when compare flags are actually set.

way2muchnoise commented:
This is starting to look at lot better than before, keep up the good work!

TeknoKit committed:
Added more framework, created button on appropriate machines, now functional

Still need to add the filter item support and test the various NBT combinations with every machine. Then consider if or how to support choosing what tags to deal with in game.

TeknoKit committed:
Remove incorrect imports

TeknoKit committed:
Added Filter support, code cleanup

All tags tested and machines working.
All left to do is decide on if/how to choose the various Forestry NBT tags in-game.

TeknoKit committed:
More code cleanup, organized imports

All buttons set the GEN and IS_ANALYZED flags for sanitation. Those tags are the most common tags most would want to ignore with autonamation in Forestry. No buttons are loaded if Forestry is not present. Decided the filter item would be best for any more complicated Forestry NBTTag filtering if desired but the Filter GUI is already quite full.

TeknoKit committed:
Update to 1.5.13, revert import changes

TeknoKit committed:
Forgot some lines, attempted to fix indentations

TeknoKit committed:
No idea why it thinks this is a conflict

TeknoKit committed:
Last attempt, erraneous tab

TeknoKit commented:
Should I just re-base the fork at this point?

TeknoKit commented:
I'll re-fork and start a new pull once this is fixed.
